src/core/lib/surface/call.c

```diff
@@ -1038,9 +1039,14 @@ static void finish_batch_completion(grpc_exec_ctx *exec_ctx, void *user_data,
 static void post_batch_completion(grpc_exec_ctx *exec_ctx,
                                   batch_control *bctl) {
   grpc_call *call = bctl->call;
+  grpc_error *error = bctl->error;
+  if (bctl->recv_final_op) {
+    GRPC_ERROR_UNREF(error);
+    error = GRPC_ERROR_NONE;
+  }
   if (bctl->is_notify_tag_closure) {
     /* unrefs bctl->error */
-    grpc_exec_ctx_sched(exec_ctx, bctl->notify_tag, bctl->error, NULL);
+    grpc_closure_run(exec_ctx, bctl->notify_tag, error);
```

Comment:
closure_run is new, and its use is replicated around a bunch in this PR. This branch ('direct-calls') gets its name from closure_run's use. There's some locality advantages from bringing the chaining together (not a big surprise). 

What was a surprise is that most uses of exec_ctx_sched were from closures themselves, which are ('by definition') safe points for executing closures. So... small win from just executing stuff. 

We get a bigger win because since combiner locks execute closures, and they ALSO execute them from safe points, by directly executing the closure, we avoid having to fiddle with the combiner lock state (by leaving it and falling back to the execution context), and so we make combiner locks a lot more of a win.

Anyway, once this code drops, the advice changes from 'always use exec_ctx_sched' to 'use closure_run whenever you're at a provable safe point in the code, and use exec_ctx_sched at all other times'.